rules/builds.smk

```diff
@@ -141,43 +141,81 @@ rule mask:
         """
 
-def _get_group_by_wildcards(wildcards):
-    region_key = f"group_by_{wildcards.region}"
-
-    if region_key in config["subsample_focus"]:
-        return config["subsample_focus"][region_key]
+def _get_subsampling_settings(wildcards):
+    # Allow users to override default subsampling with their own settings keyed
+    # by location type and name. For example, "region_europe" or
+    # "country_iceland". Otherwise, default to settings for the location type.
+    subsampling_scheme = _get_subsampling_scheme_by_build_name(wildcards.build_name)
+    subsampling_settings = config["subsampling"][subsampling_scheme]
+
+    if hasattr(wildcards, "subsample"):
+        return subsampling_settings[wildcards.subsample]
     else:
-        return config["subsample_focus"]["group_by"]
+        return subsampling_settings
+
+
+def get_priorities(wildcards):
+    subsampling_settings = _get_subsampling_settings(wildcards)
 
-def _get_sequences_per_group_by_wildcards(wildcards):
-    if wildcards.region == "global":
-        return config["subsample_focus"]["seq_per_group_global"]
+    if "priorities" in subsampling_settings and subsampling_settings["priorities"]["type"] == "proximity":
+        return f"results/{wildcards.build_name}/proximity_{subsampling_settings['priorities']['focus']}.tsv"
     else:
-        return config["subsample_focus"]["seq_per_group_regional"]
+        # TODO: find a way to make the list of input files depend on config
+        return config["files"]["include"]
 
-def _get_focus_exclude_argument_by_wildcards(wildcards):
-    if wildcards.region == "global":
-        return ""
+
+def get_priority_argument(wildcards):
+    subsampling_settings = _get_subsampling_settings(wildcards)
+
+    if "priorities" in subsampling_settings and subsampling_settings["priorities"]["type"] == "proximity":
+        return "--priority " + get_priorities(wildcards)
     else:
-        return f"--exclude-where \"region!={wildcards.region.replace('-', ' ')}\""
+        return ""
+
+
+def _get_specific_subsampling_setting(setting, optional=False):
+    def _get_setting(wildcards):
+        subsampling_scheme = _get_subsampling_scheme_by_build_name(wildcards.build_name)
+        geographic_name = _get_geographic_name_by_build_name(wildcards.build_name)
+        geographic_scale = _get_geographic_scale_by_build_name(wildcards.build_name)
+
+        if optional:
+            value = _get_subsampling_settings(wildcards).get(setting, "")
+        else:
+            value = _get_subsampling_settings(wildcards)[setting]
+
+        if isinstance(value, str):
+            value = value.format(**wildcards)
+        else:
+            return value
+
+        if ("geo_hierarchy" in config
+            and geographic_scale in config["geo_hierarchy"]
+            and geographic_name in config["geo_hierarchy"][geographic_scale]):
+            return value.format(**config["geo_hierarchy"][geographic_scale][geographic_name])
+        else:
```

Comment:
To prevent confusing errors when the geo hierarchy doesn't exist either because it wasn't provided or because the value doesn't exist in it, we should add a check here along the lines of
```
elif re.search(r'\{.+\}', value):
    raise Exception(f"The parameters for the subsampling (sub)scheme {wildcards.subsample} require expansion from the geo hierarchy but it is not defined. Check that the build params define the correct geographic name and scale, and that these are present in the `geo_hierarchy`")
```
I don't think there's ever a valid argument to `augur filter` which contains curly braces, but perhaps?!?

Reply:
P.S. the entire if/else block could be replaced with
```
return value.format(**config.get("geo_hierarchy", {}).get(geographic_scale, {}).get(geographic_name, {}))
```
But I'm in favor of keeping it verbose and adding comments & checks

Reply:
This is a great suggestion. Several builds from the `swiss` profile were silently failing for me because of this issue. Just fixed it.